### PR TITLE
Allow users to modify the core and read replica Containers before they are created

### LINKED ISSUE #43: Add support for setting the neo4j config.
I'm not sure what the best way to do this is.

We could add a `Supplier<Map<String, String>>` to the annotation:
`@NeedsCausalCluster(config=MyClass.getNeo4jConfig())`

We could add an annotation that can be added to either a method or a field on a class which contains the config

```
@NeedsCausalCluster
public class AkkaRestartTests
{
    private final Logger log = LoggerFactory.getLogger( this.getClass() );

    @CausalCluster
    private static Neo4jCluster cluster;

    @CausalClusterConfig
    private static Map<String, String> config = Map.of("causal_clustering.middleware.logging.level", "DEBUG");
```

But perhaps you Java Wizards have some better ideas @meistermeier @michael-simons ?

### PR DESCRIPTION
**Pull request by eastlondoner**
Touched a lot of file and did a few formatting tidy ups

### CODE REVIEW

File: src/main/java/org/neo4j/junit/jupiter/causal_cluster/CausalClusterExtension.java [hunk 1]
```diff
@@ -35,7 +38,9 @@
 import org.junit.jupiter.api.extension.ExtensionContext;
 import org.junit.jupiter.api.extension.TestInstances;
 import org.junit.platform.commons.support.AnnotationSupport;
-import org.junit.platform.commons.support.HierarchyTraversalMode;
+import org.testcontainers.containers.Neo4jContainer;
+
+import static org.junit.platform.commons.support.HierarchyTraversalMode.TOP_DOWN;
```
Comment (michael-simons) [inline src/main/java/org/neo4j/junit/jupiter/causal_cluster/CausalClusterExtension.java]: I wonder if that goes through my check style config… :hmm:

**Review by michael-simons — APPROVED** [review]
Looks great actually!

The only thing that makes me wonder if the modifier should be a `Consumer` instead of a function. But with the function it fits very nicely with the way test containers allow that modification.

We should probably make it an https://docs.oracle.com/javase/8/docs/api/java/util/function/UnaryOperator.html

File: src/main/java/org/neo4j/junit/jupiter/causal_cluster/CausalClusterExtension.java (hunk 1 above)
Comment (eastlondoner) [reply src/main/java/org/neo4j/junit/jupiter/causal_cluster/CausalClusterExtension.java]: it doesn't.

**Comment by eastlondoner** [comment]
I changed the interface to extend `UnaryOperator` :-) thanks for the review

**Comment by michael-simons** [comment]
Do we need an interface for it at all?

**Comment by eastlondoner** [comment]
We don't *need* one, but I think it makes the code more readable. 

The interface is purely internal, the users of this don't implement it - any lambda or method that satisfies it will work.

**Comment by michael-simons** [comment]
> any lambda or method that satisfies it will work.

So please let an operator be enough.

**Pull request merged**